main.go

```diff
@@ -81,21 +93,50 @@ func serverFinalizer(server *Server) {
 //
 // FIXME Pass in a list of 8-bit tags defining epochs.
 // The instance will generate its own secret key.
-func NewServer() (*Server, error) {
+func NewServer() (*Server, *string, error) {
 	// FIXME should we runtime.LockOSThread() here?
 	raw := C.randomness_server_create()
 	if raw == nil {
-		return nil, errors.New("failed to create randomness server")
+		return nil, nil, errors.New("failed to create randomness server")
 	}
 	server := &Server{raw: raw}
 	runtime.SetFinalizer(server, serverFinalizer)
-	return server, nil
+
+	var pkOutput [16384]byte
+	pkSize := C.randomness_server_get_public_key(
+		server.raw, (*C.uint8_t)(unsafe.Pointer(&pkOutput[0])))
+	if pkSize == 0 {
+		return nil, nil, errors.New("failed to get public key")
+	}
+	pk := base64.StdEncoding.EncodeToString(pkOutput[:pkSize])
+
+	return server, &pk, nil
+}
+
+// getEpoch returns the current epoch for the given timestamp.
+func getEpoch(t time.Time) (epoch, int64) {
+	currentMillisSinceFirstEpoch := t.UnixMilli() - firstEpochTimestampMs
+	epochsSinceFirstEpoch := currentMillisSinceFirstEpoch / epochLengthTimeMs
+	nextEpochTimeMs := firstEpochTimestampMs +
+		(epochLengthTimeMs * (epochsSinceFirstEpoch + 1))
+	currentEpoch := epochsSinceFirstEpoch % 256
+	return epoch(currentEpoch), nextEpochTimeMs
 }
 
-// getEpoch returns the ISO week number of the given timestamp.
-func getEpoch(t time.Time) epoch {
-	_, week := t.ISOWeek()
-	return epoch(week)
+// getServerInfo returns the current epoch info and public key.
+func getServerInfo(pk string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		currentEpoch, nextEpochTimeMs := getEpoch(time.Now())
+		resp := serverInfoResponse{
+			PublicKey:       pk,
```

Comment:
Can go interpolate json? Having sta-rs serialize the key struct to json instead of bincode might achieve more transparency with the same amount of code here.

I understand "It can be a single opaque buffer in my implementation," but that's a layer violation. There could be other implementations with different requirements. (Hopefully there will be if our IETF proposals are successful!) They might find the nested key encoding the opposite of convenient.

One of our goals here is auditability. json helps with that by being simple read, and base64 makes it straightforward to cut-and-paste inherently binary parts like points and keys for testing and evaluation. Having the internal structure of a binary part in a separate, poorly-specified encoding treats that part of the response as opaque.

Anyway, that's my perspective. I'm happy to address it separately, as @NullHypothesis suggested.

Reply:
I think the goal of auditability/transparency remains intact, since one may review the sta-rs/ppoprf code to see what's being encoded. However, I do think readability of the `ServerPublicKey` via the network response is impacted. A few options to address this:

- Switch to JSON encoding, like you suggested
- Create a CLI utility to parse and output decoded information for auditors (similar to what the openssl cli does for parsing X.509 certs, public keys, etc.). This could be useful for nested STAR messages as well, since the plan to use base64 + bincode for those is more or less permanent.
- Leave everything as is. Let the sta-rs/ppoprf code speak for itself

For now, I think it makes sense to leave things as is for now, since the bincode serialization change/PR in sta-rs was already merged by yourself and Alex. I think a small CLI would be kind of neat, even for debugging/development purposes